Refactors and updates the RSS reader lambda. Adds tests.

paulpilone opened the pull request.

paulpilone committed:
Updates RSS lambda. Adds files neeeded by that lambad.

paulpilone commented on lib/dspace-schemas.ts:
```diff
@@ -0,0 +1,98 @@
+import { z } from 'zod';
```
This file is a WIP but is necessary to pull into this commit because it defines schemas related to the RSS feed. I'd really like to just merge this in and know that I'm still working on it on my other branch.

paulpilone committed:
Adds types for xml2js.

paulpilone committed:
Adds missing cdk lambdas file.

paulpilone committed:
Refactors metadata-downloader.

marchuffnagle commented on lib/dspace-client.ts:
```diff
@@ -0,0 +1,193 @@
+    resolveBodyOnly: true,
+  });
+
+  return new Promise((resolve, reject) => {
```
Parser supports Promises. This can be cleaned up to:

```ts
const parser = new Parser({ explicitRoot: false });
return parser.parseStringPromise(rawFeedXML);
  ```

marchuffnagle commented on lib/dspace-client.ts:
```diff
@@ -0,0 +1,193 @@
+    });
+  });
+
+  // TODO: Validate our RSS feed with Zod.
```
Please file a GitHub issue to address this and note the issue number here in the code.

paulpilone replied:
OBE

marchuffnagle commented on lib/dspace-client.ts:
```diff
@@ -0,0 +1,193 @@
+export const getItems = async (
+  endpoint: string,
+  searchParams: GetItemsSearchParams = {}
+): Promise<DSpaceItem[]> => {
```
We need to be validating this response if we're asserting its type.

marchuffnagle commented on lib/dspace-client.ts:
```diff
@@ -0,0 +1,193 @@
+export const getItem = async (
+  endpoint: string,
+  uuid: string
+): Promise<DSpaceItem | undefined> => {
```
We need to be validating this response if we're asserting its type.

marchuffnagle commented on lib/dspace-client.ts:
```diff
@@ -0,0 +1,193 @@
+export const getMetadata = async (
+  endpoint: string,
+  uuid: string
+): Promise<Metadata[] | undefined> => {
```
We need to be validating this response if we're asserting its type.

marchuffnagle commented on ingest/lambdas/rss-feed-ingester.js:
```diff
@@ -0,0 +1,92 @@
+const { getStringFromEnv } = require('../../lib/env-utils');
+
+const dspaceFeedReadInterval = process.env.DSPACE_FEED_READ_INTERVAL
+  ? Number.parseInt(process.env.DSPACE_FEED_READ_INTERVAL)
```
It may be worth adding a `getIntFromEnv` function to `env-utils`

marchuffnagle commented on ingest/lambdas/rss-feed-ingester.js:
```diff
@@ -0,0 +1,92 @@
+        const dspaceItems = await dspaceClient.find(
+          dspaceEndpoint,
+          'dc.identifier.uri',
+          feedItem.link[0]
```
Could this potentially be undefined (an empty array)?

paulpilone replied:
See my comment about the updated RSS schema parsing.

marchuffnagle commented on ingest/lib/ingest-queue.ts:
```diff
@@ -0,0 +1,31 @@
+ * @returns Returns the resut of the SNS PublishCommand:
+ * https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-sns/interfaces/publishcommandoutput.html
+ */
+export const queueIngestDocument = async (
```
There's not really anything specific to ingest in this function. If you changed the argument to `message` instead of `uuid`, you could put it in an `sns-utils` function. At that point, though, there isn't really anything beyond [`publi](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/SNS.html#publish-property)

paulpilone replied:
My thought here was that SNS isn't really the best thing to use for ingest and it should be updated (I'd like to create an Issue for this) and if we did fix that we could just make that change here instead of fixing everything that queues up a UUID for ingest 🤷

paulpilone replied:
https://github.com/iodepo/OceanBestPractices/pull/106/files#r762453817

Even though `client.send()` is async?

marchuffnagle replied:
> Even though `client.send()` is async?

Correct. `client.send()` returns a `Promise` but, because you aren't using `await` to extract the value that the promise resolves to, and are returning that promise directly, there's no need for `async` here.

marchuffnagle commented on ingest/lambdas/rss-feed-ingester.js:
```diff
@@ -0,0 +1,92 @@
+    // dc.identifier.uri field since that's all that is available in the feed.
+    // This identifier doesn't have a consistent domain so we use the UUID for
+    // ingest queuing.
+    await pMap(
```
Consider creating `lib/sns-utils.ts` with:

```ts
import { chunk } from 'lodash';
import pMap from 'p-map';
import { sns } from './aws-clients';

interface PublishBatchInput {
  topicArn: string
  messages: string[],
  concurrency?: number
}

export const publishBatch = async (
  params: PublishBatchInput
): Promise<void> => {
  const { concurrency, messages, topicArn: TopicArn } = params;

  await pMap(
    chunk(messages, 10),
    (messageChunk) => {
      const PublishBatchRequestEntries = messageChunk.map((Message, index) => ({
        Id: index.toString(),
        Message,
      }));

      return sns().publishBatch({
        TopicArn,
        PublishBatchRequestEntries,
      }).promise();
    },
    { concurrency }
  );
};
```

With that utility, this could be refactored to something like:

```ts
  const dSpaceItems = await pMap(
      feed.channel[0].item,
      (feedItem) => dspaceClient.find(
        dspaceEndpoint,
        'dc.identifier.uri',
        feedItem.link[0]
      ),
      { concurrency: 5 }
    );

    const uuids = dSpaceItems.flat().map((i) => i.uuid);

    await publishBatch({
      topicArn: ingestTopicArn,
      messages: uuids,
      concurrency: 5,
    });
```

That feels a bit easier to read.

paulpilone replied:
@marchuffnagle Could we file this as a Github issue? I think this is a good idea and I'd like to get it working with Localstack, etc. but we'd have to make sure the tests are updated and and configure the SNS client in `aws-clients` – all good things but considering the RSS reader is fairly trivial code (and we're only ever dealing with at most 5 items in the feed anyway) maybe it's better to circle back around to this later?

marchuffnagle replied:
> Could we file this as a GitHub issue?

Sure.

paulpilone commented on ingest/lambdas/rss-feed-ingester.js:
```diff
@@ -0,0 +1,92 @@
+ * @returns {boolean} Whether or not we should publish new documents based
+ *          on the last time we checked the RSS feed
+ */
+const shouldQueuePublishedDocuments = (publishedDate, feedReadInterval) => {
```
Changed it to `Date.now() - feedReadInterval` and added a comment. I'll file an Issue in Github though because we shouldn't be relying on the read interval but instead save off the "last read date" somewhere and use that for this check.

paulpilone commented on ingest/lambdas/rss-feed-ingester.js:
```diff
@@ -0,0 +1,92 @@
+
+    // Parse out the last published date and use it to determine whether or
+    // not we should index new documents.
+    const lastPublishedDate = new Date(feed.channel[0].pubDate[0]._);
```
I've updated the RSS Schema to force the channel, pubDate, and items array to be nonempty. We're parsing that now in the DSpace client code so if they are empty it'll throw an error that we catch.

paulpilone commented on ingest/lambdas/rss-feed-ingester.js:
```diff
@@ -0,0 +1,92 @@
+
+        if (dspaceItems.length > 0) {
+          const dspaceItem = dspaceItems[0];
+          await ingestQueue.queueIngestDocument(
```
We can but I'm not super worried about it - the RSS feed has at most 5 items in it so it's not a huge problem to just issue these requests individually.

paulpilone committed:
Addresses PR feedback.

paulpilone commented:
@marchuffnagle I've addressed most of your feedback in the latest commit. I responded to a few of them and if you feel strongly about addressing them before this is approved I understand that. I do think some of those - given the not-so-complex nature of this function could be addressed in a separate issue later on.

paulpilone committed:
Merge branch 'rss-feed-ingester-refactor' into metadata-downloader-refactor

paulpilone committed:
Improves metadata-downloader tests.

marchuffnagle commented on ingest/lambdas/rss-feed-ingester.js:
```diff
@@ -0,0 +1,89 @@
+
+  // Assuming the last time we checked for an updated feed was feedReadInterval
+  // seconds ago, if the published date is newer than that we return true.
+  const feedReadIntervalDate = Date.now() - feedReadInterval;
```
`Operator '>=' cannot be applied to types 'Date' and 'number'.ts(2365)`

paulpilone replied:
Fixed (I think) but why don't I see that compiler error?

marchuffnagle commented on lib/dspace-schemas.ts:
```diff
@@ -0,0 +1,75 @@
+
+export type Metadata = z.infer<typeof metadataSchema>;
+
+// export interface Metadata {
```
Can the commented-out interfaces in this file be deleted?

marchuffnagle commented on lib/dspace-schemas.ts:
```diff
@@ -0,0 +1,75 @@
+  uuid: z.string().uuid(),
+  handle: z.string(),
+  lastModified: z.string(),
+  bitstreams: z.array(bitstreamSchema),
```
I really like this schema reuse.

paulpilone committed:
Fixes incorrect date comparison.

Commit:
Updates from pairing

marchuffnagle reviewed (approved):
👍

Commit:
Merge pull request #107 from iodepo/metadata-downloader-refactor
